src/dune_pkg/opam_repo.ml

```diff
@@ -265,59 +274,110 @@ let load_opam_package t opam_package =
+         ]
+     | Ok version_dirs -> List.map version_dirs ~f:OpamPackage.of_string)
+;;
+
+let all_packages_versions_at_rev rev opam_package_name =
+  let version_dir_path =
+    let name = OpamPackage.Name.to_string opam_package_name in
+    Path.Local.relative (Path.Local.of_string "packages") name
+  in
+  Rev_store.Remote.At_rev.directory_entries rev version_dir_path
+  |> Rev_store.File.Set.to_list
+  |> List.filter_map ~f:(fun file ->
+    let path = Rev_store.File.path file in
+    let open Option.O in
+    Path.Local.basename_opt path
+    >>= function
+    | "opam" ->
+      let+ package =
+        Path.Local.parent path >>| Path.Local.basename >>| OpamPackage.of_string
+      in
+      file, package
+    | _ -> None)
+;;
+
 let all_package_versions t opam_package_name =
   match t.source with
-  | Directory d ->
-    (match get_opam_package_version_dir_path d opam_package_name with
-     | None -> []
-     | Some version_dir_path ->
-       (match Path.readdir_unsorted version_dir_path with
-        | Error e ->
-          User_error.raise
-            [ Pp.textf
-                "Unable to read package versions from %s: %s"
-                (Path.to_string_maybe_quoted version_dir_path)
-                (Dune_filesystem_stubs.Unix_error.Detailed.to_string_hum e)
-            ]
-        | Ok version_dirs -> List.map version_dirs ~f:OpamPackage.of_string))
-  | Repo at_rev ->
-    let version_dir_path =
-      let name = OpamPackage.Name.to_string opam_package_name in
-      Path.Local.relative (Path.Local.of_string "packages") name
-    in
-    Rev_store.Remote.At_rev.directory_entries at_rev version_dir_path
-    |> Path.Local.Set.to_list
-    |> List.filter_map ~f:(fun dir_entry ->
-      let open Option.O in
-      Path.Local.basename_opt dir_entry
-      >>= function
-      | "opam" ->
-        let+ parent = Path.Local.parent dir_entry in
-        parent |> Path.Local.basename |> OpamPackage.of_string
-      | _ -> None)
+  | Directory dir ->
+    all_packages_versions_in_dir ~dir opam_package_name
+    |> List.map ~f:(fun pkg -> None, pkg)
+  | Repo rev ->
+    all_packages_versions_at_rev rev opam_package_name
+    |> List.map ~f:(fun (file, pkg) -> Some file, pkg)
 ;;
 
 let load_all_versions ts opam_package_name =
-  List.map ts ~f:(fun t ->
-    all_package_versions t opam_package_name |> List.rev_map ~f:(fun pkg -> t, pkg))
-  |> List.concat
-  |> List.fold_left ~init:OpamPackage.Version.Map.empty ~f:(fun acc (repo, package) ->
-    let version = OpamPackage.version package in
-    if OpamPackage.Version.Map.mem version acc
-    then acc
-    else OpamPackage.Version.Map.add version (repo, package) acc)
-  |> OpamPackage.Version.Map.to_seq
-  |> List.of_seq
-  |> Fiber.parallel_map ~f:(fun (version, (repo, pkg)) ->
-    load_opam_package repo pkg
-    >>| Option.map ~f:(fun (pkg : With_file.t) -> version, (repo, pkg)))
-  >>| List.filter_opt
-  >>| OpamPackage.Version.Map.of_list
+  let from_git, from_dirs =
```

Comment:
I do not understand what the reason is that the `all_package_versions` are to be distinguished between git and directory and require to be handled differently in `load_all_versions`. Why this particular choice?

Reply:
Because fetching from git can be batched into a single process execution, while fetching from directories cannot. This code is just preparing the batch.